File: main/lilys-homework/lilys-homework.pl
#!/usr/bin/env perl

use strict;
use warnings;
use feature qw(say);
use List::Util qw(min);

use constant VISITED => -1;

sub normalize_to_permutation {
    my ($values) = @_;

    my @permutation = sort { $values->[$a] <=> $values->[$b] }
                           (0..$#{$values});
    return @permutation;
}

# Traverses the cycle of the permutation that begins at the index start,
# marking each element visited and returning the number of positions traversed
# (always at least 1).
sub traverse_and_mark_cycle {
    my ($permutation, $start) = @_;

    my $i = $start;
    my $cycle_length = 1;

    while (1) {
        # advance to the next value in the cycle
        ($permutation->[$i], $i) = (VISITED, $permutation->[$i]);

        if ($i == $start) { return $cycle_length }
        ++$cycle_length;
    }
}

sub compute_min_swaps {
    my ($values) = @_;

    my @permutation = normalize_to_permutation $values;
    my $swaps = 0;

    for my $i (0..$#permutation) {
        if ($permutation[$i] != VISITED) {
            $swaps += traverse_and_mark_cycle(\@permutation, $i) - 1;
        }
    }

    return $swaps;
}

<>; # don't need n
my @values = split /\s+/msx, <>;
say min(compute_min_swaps(\@values), compute_min_swaps([reverse @values]));

File: main/lilys-homework/lilys-homework.py
#!/usr/bin/env pypy3

import itertools

VISITED = -1

def normalize_to_permutation(values, reverse):
    return sorted(range(len(values)), reverse=reverse, key=lambda i: values[i])

def traverse_and_mark_cycle(permutation, start):
    '''Traverses the cycle of the permutation that begins at the index start,
    marking each element visited and returning the number of positions
    traversed (always at least 1).'''
    i = start
    for cycle_length in itertools.count(1):
        permutation[i], i = VISITED, permutation[i] # advance to next value
        if i == start:
            return cycle_length

def compute_min_swaps(values, reverse):
    permutation = normalize_to_permutation(values, reverse)

    swaps = 0
    for i in range(len(permutation)):
        if permutation[i] != VISITED:
            swaps += traverse_and_mark_cycle(permutation, i) - 1

    return swaps

input() # don't need n
values = list(map(int, input().split()))

print(min(compute_min_swaps(values, False), compute_min_swaps(values, True)))
